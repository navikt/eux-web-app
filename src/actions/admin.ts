import {ActionWithPayload, call} from "@navikt/fetch";
import * as types from 'constants/actionTypes'
import * as urls from 'constants/urls'
// @ts-ignore
import { sprintf } from 'sprintf-js';

export const publishSedEvents = (
  sedEvents: Array<{rinasakId: string, documentId: string, documentVersion: string}>
): ActionWithPayload<any> => {
  return call({
    url: sprintf(urls.ADMIN_PUBLISH_SED_EVENTS_URL),
    cascadeFailureError: true,
    expectedPayload: {
      "userMessage": "Følgende SED ble ikke publisert: 1751447_27b266cefb294c81aced7d0100a56cab_1",
      "detail": "Sjekk detaljer for hver hendelse for feilårsaker",
      "successCount": 1,
      "failureCount": 1
    },
    expectedErrorRate: {
      200 : 0.3, 207: 0.3, 500: 0.4
    },
    method: 'POST',
    body: {
      sedEvents
    },
    type: {
      request: types.ADMIN_PUBLISH_SED_EVENTS_REQUEST,
      success: types.ADMIN_PUBLISH_SED_EVENTS_SUCCESS,
      failure: types.ADMIN_PUBLISH_SED_EVENTS_FAILURE
    }
  })
}

export const adminResetStatus = () => ({
  type: types.ADMIN_RESET_STATUS,
})

