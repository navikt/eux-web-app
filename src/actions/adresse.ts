import * as types from 'constants/actionTypes'
import * as urls from 'constants/urls'
import { Adresse } from 'declarations/sed'
import { ActionWithPayload, call } from '@navikt/fetch'
import mockAdresse from 'mocks/adresse/adresse'
import { Action, ActionCreator } from 'redux'
const sprintf = require('sprintf-js').sprintf

export const resetAdresse: ActionCreator<Action> = (): Action => ({
  type: types.ADRESSE_SEARCH_RESET
})

export const searchAdresse = (
  fnr: string
): ActionWithPayload<Array<Adresse>> => {
  return call({
    url: sprintf(urls.API_ADRESSE_URL, { fnr }),
    expectedPayload: mockAdresse,
    cascadeFailureError: true,
    type: {
      request: types.ADRESSE_SEARCH_REQUEST,
      success: types.ADRESSE_SEARCH_SUCCESS,
      failure: types.ADRESSE_SEARCH_FAILURE
    }
  })
}
