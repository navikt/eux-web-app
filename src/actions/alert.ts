import * as types from 'constants/actionTypes'
import { ActionWithPayload } from '@navikt/fetch'
import { Action, ActionCreator } from 'redux'

export const alertClear: ActionCreator<Action> = (): Action => ({
  type: types.ALERT_CLEAR
})

export const alertFailure: ActionCreator<ActionWithPayload<{message: JSX.Element | string}>> = (
  message: JSX.Element | string
): ActionWithPayload<{message: JSX.Element | string}> => ({
  type: types.ALERT_FAILURE,
  payload: {
    message
  }
})

export const alertSuccess: ActionCreator<ActionWithPayload<{message: JSX.Element | string}>> = (
  message: JSX.Element | string
): ActionWithPayload<{message: JSX.Element | string}> => ({
  type: types.ALERT_SUCCESS,
  payload: {
    message
  }
})

export const alertWarning: ActionCreator<ActionWithPayload<{message: JSX.Element | string}>> = (
  message: JSX.Element | string
): ActionWithPayload<{message: JSX.Element | string}> => ({
  type: types.ALERT_WARNING,
  payload: {
    message
  }
})
