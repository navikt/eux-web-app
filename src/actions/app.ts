import * as types from 'constants/actionTypes'
import * as urls from 'constants/urls'
import { ParamPayload } from 'declarations/app'
import { Enheter, LogMeAgainPayload, Saksbehandler, ServerInfo, UtgaarDatoPayload } from 'declarations/types'
import EKV from '@navikt/eessi-kodeverk'
import { ActionWithPayload, call } from '@navikt/fetch'
import mockEnhet from 'mocks/app/enhet'
import mockReautorisering from 'mocks/app/reautorisering'
import mockSaksbehandler from 'mocks/app/saksbehandler'
import mockServerInfo from 'mocks/app/serverinfo'
import mockUtgaarDato from 'mocks/app/utgaarDato'
import { Action, ActionCreator } from 'redux'

export const copyToClipboard = (text?: string) => ({
  type: types.APP_CLIPBOARD_COPY,
  payload: text
})

export const getEnheter = (): ActionWithPayload<Enheter> => {
  return call({
    url: urls.API_ENHETER_URL,
    expectedPayload: mockEnhet,
    type: {
      request: types.APP_ENHETER_REQUEST,
      success: types.APP_ENHETER_SUCCESS,
      failure: types.APP_ENHETER_FAILURE
    }
  })
}

export const logMeAgain = (name ?: string): ActionWithPayload<LogMeAgainPayload> => {
  // origin: http://{host:port} pathname: /pdu1, no hash
  let redirectUrl = (window.location as any).origin + (window.location as any).pathname
  if (name) {
    redirectUrl += '?name=' + name
  }

  return call({
    url: urls.API_REAUTENTISERING_URL,
    expectedPayload: mockReautorisering,
    context: {
      redirectUrl
    },
    type: {
      request: types.APP_LOGMEAGAIN_REQUEST,
      success: types.APP_LOGMEAGAIN_SUCCESS,
      failure: types.APP_LOGMEAGAIN_FAILURE
    }
  })
}

export const setStatusParam: ActionCreator<ActionWithPayload<ParamPayload>> = (
  key: string,
  value: any
): ActionWithPayload<ParamPayload> => ({
  type: types.APP_PARAM_SET,
  payload: {
    key,
    value
  } as ParamPayload
})

/*
KTObjects: {
  buctyper: buctyper,
  familierelasjoner: familierelasjoner,
  kjoenn: kjoenn,
  landkoder: landkoder,
  sektor: sektor,
  sedtyper: sedtyper,
  tema: tema
};
*/
export const preload: ActionCreator<ActionWithPayload<any>> = (
): ActionWithPayload<any> => ({
  type: types.APP_PRELOAD,
  payload: {
    ...(EKV as any).KTObjects,
    kodemaps: {
      ...(EKV as any).Kodemaps
    }
  } // kodemaps: { BUC2SEDS, SEKTOR2FAGSAK, SEKTOR2BUC }
})

export const appReset: ActionCreator<Action> = (): Action => ({
  type: types.APP_RESET
})

export const getSaksbehandler = (): ActionWithPayload<Saksbehandler> => {
  return call({
    url: urls.API_SAKSBEHANDLER_URL,
    expectedPayload: mockSaksbehandler(),
    type: {
      request: types.APP_SAKSBEHANDLER_REQUEST,
      success: types.APP_SAKSBEHANDLER_SUCCESS,
      failure: types.APP_SAKSBEHANDLER_FAILURE
    }
  })
}

export const getServerinfo = (): ActionWithPayload<ServerInfo> => {
  return call({
    url: urls.API_SERVERINFO_URL,
    expectedPayload: mockServerInfo(),
    type: {
      request: types.APP_SERVERINFO_REQUEST,
      success: types.APP_SERVERINFO_SUCCESS,
      failure: types.APP_SERVERINFO_FAILURE
    }
  })
}

export const getUtgaarDato = (): ActionWithPayload<UtgaarDatoPayload> => {
  return call({
    url: urls.API_UTGAARDATO_URL,
    expectedPayload: mockUtgaarDato,
    type: {
      request: types.APP_UTGAARDATO_REQUEST,
      success: types.APP_UTGAARDATO_SUCCESS,
      failure: types.APP_UTGAARDATO_FAILURE
    }
  })
}

export const reduceSessionTime: ActionCreator<ActionWithPayload> = (): ActionWithPayload => ({
  type: types.APP_SESSION_SET,
  payload: {
    minutes: 6
  }
})
