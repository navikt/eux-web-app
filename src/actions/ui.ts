import * as types from '../constants/actionTypes'
import { ModalContent } from '../declarations/components'
import { ActionWithPayload } from 'js-fetch-api'
import { Action } from 'redux'

export const openModal = (modal: ModalContent): ActionWithPayload<ModalContent> => ({
  type: types.UI_MODAL_SET,
  payload: modal
})

export const closeModal = (): ActionWithPayload<undefined> => ({
  type: types.UI_MODAL_SET,
  payload: undefined
})

export const toggleHighContrast = (): Action => ({
  type: types.UI_HIGHCONTRAST_TOGGLE
})
