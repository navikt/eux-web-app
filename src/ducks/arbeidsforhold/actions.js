
/**
 * Actions
 * -----------------------------------------------------------------------
 * Dette er action creators som returnerer Redux-klargjorte actions
 * uten support for asynkrone kall.
 */
