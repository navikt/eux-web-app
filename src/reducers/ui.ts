import * as types from 'constants/actionTypes'
import { ModalContent } from 'declarations/components'
import { ActionWithPayload } from '@navikt/fetch'
import { Action } from 'redux'

export interface UiState {
  highContrast: boolean
  modal: ModalContent | undefined
}

export const initialUiState: UiState = {
  highContrast: false,
  modal: undefined
}

const uiReducer = (state: UiState = initialUiState, action: Action | ActionWithPayload = { type: '' }): UiState => {
  switch (action.type) {
    case types.UI_MODAL_SET:
      return {
        ...state,
        modal: (action as ActionWithPayload).payload
      }

    case types.UI_HIGHCONTRAST_TOGGLE :

      return {
        ...state,
        highContrast: !state.highContrast
      }

    default:
      return state
  }
}

export default uiReducer
