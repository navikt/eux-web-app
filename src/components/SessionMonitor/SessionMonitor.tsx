import { logMeAgain, reduceSessionTime } from 'actions/app'
import { PDU1 } from 'declarations/pd'
import { ReplySed } from 'declarations/sed'
import PT from 'prop-types'
import React, {CSSProperties, useEffect, useState} from 'react'
import { useTranslation } from 'react-i18next'
import { useAppDispatch, useAppSelector } from 'store'
import styled from 'styled-components'
import Modal from 'components/Modal/Modal'
import { BodyLong, Button } from '@navikt/ds-react'
import { State } from 'declarations/reducers'
import _ from 'lodash'
import { IS_DEVELOPMENT, IS_Q } from 'constants/environment'
import {API_REAUTENTISERING_URL, API_UTGAARDATO_URL} from 'constants/urls'
import app, {initialAppState} from "../../reducers/app";
import * as types from "../../constants/actionTypes";
import { setTimeout, setInterval } from 'worker-timers';

const SessionMonitorDiv = styled.div`
font-size: 80%;
`

export interface SessionMonitorProps {
  checkInterval?: number
  expirationTime?: number
  sessionEndsAt?: number
  millisecondsForWarning?: number
  sessionExpiredReload?: number
  sessionAutoRenew?: number
  now?: Date
}

export interface SessionMonitorSelector {
  pdu1: PDU1 | null | undefined
  replySed: ReplySed | null | undefined
  state: State | null | undefined
}

export interface WonderwallResponse {
  session: {
    created_at: string
    ends_at: string
    timeout_at: string
    ends_in_seconds: number
    active: true
    timeout_in_seconds: number
  }
  tokens: {
    expire_at: string
    refreshed_at: string
    expire_in_seconds: number
    next_auto_refresh_in_seconds: number
    refresh_cooldown: boolean
    refresh_cooldown_seconds: number
  }
}

const mapState = (state: State): SessionMonitorSelector => ({
  pdu1: state.pdu1.pdu1,
  replySed: state.svarsed.replySed,
  state: state
})

function extractTime(response: Response, wonderwallResponse: WonderwallResponse, state: State | null | undefined) {
  let timeTuple: [number, number];
  timeTuple = [-1, -1]
  if (response.ok) {
    const tokens = wonderwallResponse?.tokens
    if (tokens) {
      const nowDate: Date = new Date()
      const diffMillis: number = new Date(tokens.expire_at).getTime() - nowDate.getTime()
      const expirationTime = new Date(tokens.expire_at).getTime()

      const session = wonderwallResponse?.session

      const sessionEndsAt: number = (session) ? new Date(session.ends_at).getTime()  : 0;
//      const expirationTime = new Date().setTime(new Date().getTime() + diffMillis)

      timeTuple = [expirationTime, sessionEndsAt];

      const diffMinutes: number = Math.ceil(diffMillis / 1000 / 60);
      console.log('Wonderwall minutes left', diffMinutes)

      if (state && state.app) {
        console.log('Wonderwall setting', diffMinutes)
        console.log('Wonderwall expirationTime', expirationTime)

        state.app.expirationTime = expirationTime
        state.app.sessionEndsAt = sessionEndsAt;
      }
    } else {
      console.log('No content')
    }
  } else {
    console.log('Failed call')

  }
  return timeTuple
}

const SessionMonitor: React.FC<SessionMonitorProps> = ({
  /* check every minute */
  checkInterval = 1000 * 60,
  /* When token will expire */
  expirationTime,
  /* When session will expire */
  sessionEndsAt,
  /* Warnings should start under 5 minutes */
  millisecondsForWarning = 5 * 1000 * 60,
  /* Reload under a minute */
  sessionExpiredReload = 1000,
  /* Automatically try to renew session in background under 30 minutes */
  sessionAutoRenew = 70 * 1000 * 60,
  now
}: SessionMonitorProps): JSX.Element => {
  const [diff, setDiff] = useState<number>(0)
  const [sessionDiff, setSessionDiff] = useState<number>(0)
  const [modal, setModal] = useState<boolean>(false)
  const { pdu1, replySed, state }: SessionMonitorSelector = useAppSelector(mapState)

  const { t } = useTranslation()
  const dispatch = useAppDispatch()

  const getDiff = (expirationTime: number, now: any) => {
    const _now: Date = new Date()
    console.log('expirationTime', expirationTime)
    console.log('now', _now.getTime())
    const diff: number = expirationTime - _now.getTime()
    console.log('minutes left', Math.ceil(diff / 1000 / 60))
    setDiff(diff)
    return diff
  }

  const getSessionDiff = (expirationTime: number, now: any) => {
    const _now: Date = new Date()
    console.log('sessionEndsAt', expirationTime)
    console.log('now', _now.getTime())
    const diff: number = expirationTime - _now.getTime()
    console.log('minutes left', Math.ceil(diff / 1000 / 60))
    setSessionDiff(diff)
    return diff
  }

  const triggerReload = () => {
    window.location.reload()
  }

  async function checkTimeout () {
    let wonderwallTimeout = await currentWonderwallTimeout()
    console.log('currentWonderwallTimeout', wonderwallTimeout)
    let tokenExpirationTime = wonderwallTimeout[0];
    let sessionEndTime = wonderwallTimeout[1];
    if (!_.isNumber(tokenExpirationTime)) {
      return
    }
    const diff = getDiff(tokenExpirationTime, now)
    console.log('diff', diff)
    const sessionDiff = getSessionDiff(sessionEndTime, now)
    console.log('sessionDiff', sessionDiff)

    if (diff < sessionExpiredReload) {
      console.log('trigger reload')
      triggerReload()
    }
    if (diff < millisecondsForWarning) {
      console.log('trigger modal')
      setModal(true)
    }
    if (diff < sessionAutoRenew) {
      if (tokenExpirationTime < (sessionEndTime - 1000)) {
        console.log('trigger checkWonderwallTimeout')
        checkWonderwallTimeout()
      } else {
        console.log('trigger nothing. session max')
      }
    }
  }

  function checkTimeoutWrapper() {
    checkTimeout()
    return 0
  }

  async function checkWonderwallTimeout() {
    const response = await fetch(API_REAUTENTISERING_URL,  {
      method: "POST"
    })
    const wonderwallResponse: WonderwallResponse = await response.json()
    return extractTime(response, wonderwallResponse, state);

  }

  async function currentWonderwallTimeout() { // return tuple
    const response = await fetch(API_UTGAARDATO_URL,  {
      method: "GET"
    })
    const wonderwallResponse: WonderwallResponse = await response.json()
    return extractTime(response, wonderwallResponse, state)
  }

  useEffect(() => {
    if (expirationTime !== undefined) {
      getDiff(expirationTime, now)
      if (sessionEndsAt !== undefined) {
        getSessionDiff(sessionEndsAt, now)
      }
      setInterval(checkTimeout, checkInterval)
    }
  }, [expirationTime, sessionEndsAt])

  const title = t(diff > millisecondsForWarning ? 'app:session-ok-title' : 'app:session-expire-title')
  const text = []
  text.push(t(diff > millisecondsForWarning ? 'app:session-ok-text' : 'app:session-expire-text', { minutes: Math.ceil(diff / 1000 / 60) }))
  const hasDraft = !_.isNil(pdu1) || !_.isNil(replySed)

  if (hasDraft) {
    text.push(t('app:session-you-have-draft', { utkast: !_.isNil(pdu1) ? 'PD U1' : 'Svar SED' }))
  }

  const modalButtons = [
    {
      main: true,
      text: t('el:button-ok-got-it'),
      onClick: () => setModal(false)
    }, {
      main: false,
      text: t('el:button-log-me-again'),
      onClick: () => dispatch(logMeAgain())
    }
  ]

  if (IS_Q || IS_DEVELOPMENT) {
    modalButtons.splice(2, 0, {
      main: false,
      text: t('app:session-reduce-to-6-min'),
      onClick: () => dispatch(reduceSessionTime())
    })
  }

  return (
    <SessionMonitorDiv>
      <Modal
        open={modal}
        onModalClose={() => setModal(false)}
        modal={{
          modalTitle: title,
          modalText: (<>{text.map(t => <BodyLong key={t}>{t}</BodyLong>)}</>),
          modalButtons
        }}
      />
      <Button variant='tertiary' size='small' onClick={() => setModal(true)}>
        Token utløper om {Math.ceil(diff / 1000 / 60)} min.
        { sessionDiff != null && (sessionDiff < 380) ? <span color="red">Sesjon utløper om {Math.ceil(sessionDiff / 1000 / 60)} min</span> : ''}

      </Button>
    </SessionMonitorDiv>
  )
}

SessionMonitor.propTypes = {
  checkInterval: PT.number,
  expirationTime: PT.number,
  sessionEndsAt: PT.number,
  millisecondsForWarning: PT.number,
  sessionExpiredReload: PT.number,
  sessionAutoRenew: PT.number,
  now: PT.instanceOf(Date)
}
export default SessionMonitor
