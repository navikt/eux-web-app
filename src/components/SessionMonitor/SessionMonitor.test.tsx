import { openModal } from 'actions/ui'
import { mount, ReactWrapper } from 'enzyme'
import React from 'react'
import SessionMonitor, { SessionMonitorProps } from './SessionMonitor'

jest.mock('actions/ui', () => ({
  openModal: jest.fn(),
  closeModal: jest.fn()
}))

describe('components/SessionMonitor', () => {
  let wrapper: ReactWrapper
  const initialMockProps: SessionMonitorProps = {
    expirationTime: new Date(2020, 1, 1)
  }

  it('Renders', () => {
    const aDate = new Date('2020-12-17T03:24:00')
    const expirationTime = new Date('2020-12-17T03:24:10')
    wrapper = mount(
      <SessionMonitor
        now={aDate}
        expirationTime={expirationTime}
        checkInterval={500}
        millisecondsForWarning={9900}
        sessionExpiredReload={1000}
        {...initialMockProps}
      />)
    expect(wrapper.isEmptyRender()).toBeFalsy()
    expect(wrapper).toMatchSnapshot()
  })

  it('SessionMonitor will trigger a openModal when session is almost expiring', async (done) => {
    // expires in 5 seconds - will check every 0.5s - warnings start at 9.9s - reload only happens under 1s
    const aDate = new Date('2020-12-17T03:24:00')
    const expirationTime = new Date('2020-12-17T03:24:05')
    wrapper = mount(
      <SessionMonitor
        now={aDate}
        expirationTime={expirationTime}
        checkInterval={500}
        millisecondsForWarning={9900}
        sessionExpiredReload={1000}
        {...initialMockProps}
      />)
    expect(openModal).not.toHaveBeenCalled()
    await new Promise(resolve => {
      setTimeout(() => {
        expect(openModal).toHaveBeenCalled()
        done()
        resolve()
      }, 1000)
    })
  })

  it('SessionMonitor will trigger a openModal when session expired', async (done) => {
    // expires in 1 seconds - will check every 0.5s - warnings start at 0.9s - reload happens under 10s
    (window.location.reload as jest.Mock).mockReset()
    const aDate = new Date('2020-12-17T03:24:00')
    const expirationTime = new Date('2020-12-17T03:23:59')
    wrapper = mount(
      <SessionMonitor
        now={aDate}
        expirationTime={expirationTime}
        checkInterval={500}
        millisecondsForWarning={900}
        sessionExpiredReload={10000}
        {...initialMockProps}
      />)
    await new Promise(resolve => {
      setTimeout(() => {
        expect(window.location.reload).toHaveBeenCalled();
        (window.location.reload as jest.Mock).mockRestore()
        done()
        resolve()
      }, 1000)
    })
  })
})
