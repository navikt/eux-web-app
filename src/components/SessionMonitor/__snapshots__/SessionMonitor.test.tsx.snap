// Jest Snapshot v1, https://goo.gl/fbAQLP

exports['components/SessionMonitor Renders 1'] = `
<SessionMonitor
  checkInterval={500}
  expirationTime={2020-01-31T23:00:00.000Z}
  millisecondsForWarning={9900}
  now={2020-12-17T02:24:00.000Z}
  sessionExpiredReload={1000}
>
  <styled.div>
    <div
      className="sc-AxjAm kmXdIo"
    >
      -461004 min
    </div>
  </styled.div>
</SessionMonitor>
`
